### PR TITLE
(feat): Model API design

### PR DESCRIPTION
**Pull request by FFFiend**
**What kind of change does this PR introduce?**
Added a rough sketch of what the infrastructure surrounding the models we use might look like. Addresses #252 , and #368 and may also be tangentially/directly relevant to other issues

**Checklist**
* [ ] My code follows the style guidelines of OpenAdapt
* [ ] I have performed a self-review of my code
* [ ] If applicable, I have added tests to prove my fix is functional/effective
* [ ] I have linted my code locally prior to submission
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [ ] New and existing unit tests pass locally with my changes

**How can your code be run and tested?**

**Other information**

### CODE REVIEW

File: openadapt/ml_models/ml_models.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+PRIORITY = ["CPU", "GPU", "HOSTED"]
+from loguru import logger
```
Comment (0dm) [inline openadapt/ml_models/ml_models.py:2]:
```suggestion
from loguru import logger

PRIORITY = ["CPU", "GPU", "HOSTED"]
```

**Comment by abrichr** [comment]
Thanks for getting started on this @FFFiend !

What do you think about:

```
class Modality:
    TEXT = "TEXT"
    IMAGE = "IMAGE"

MODALITY_BY_DB_MODEL = {
    models.ActionEvent: Modality.TEXT,
    models.Screenshot: Modality.IMAGE,
    models.WindowEvent: Modality.TEXT,
}

ML_MODELS = [
    {
        "name": "gpt-4",
        "provider": openai.ml.openai,
        # "tune_func": openadapt.ml.openai.gpt4.tune,
        # "infer_func": openadapt.ml.openai.gpt4.infer,
        "modalities": openadapt.ml.openai.gpt4.MODALITIES,  #  gpt4/__init__.py: MODALITIES =  [Modality.TEXT],
    },
     {
        "name": "BLIP2",
        "provider": openai.ml.huggingface,
        # "tune_func": openadapt.ml.huggingface.blip2.tune,
        # "infer_func": openadapt.ml.huggingface.blip2.infer,
        "modalities": openadapt.ml.huggingface.blip2.MODALITIES,  #  huggingface/__init__.py: MODALITIES =  [Modality.TEXT, Modality.IMAGE],
    },
    ...
]

    
def select_ml_model(recording_ids):
    recording_data = get_data_by_modality(recording_ids)
    for ml_model in ML_MODELS:
        # evaluate the performance of the given models autoregressively on an entire recording (optionally with start_timestamp/end_timestamp parameters for evaluating on a section of a recording)
```

**Comment by abrichr** [comment]
@FFFiend @LaPetiteSouris please watch: https://drive.google.com/file/d/1UJ8PbhxnoY82-oFT786ZPw_zQY7-P5JH/view

See also: https://mldsai.slack.com/archives/C050869D3K4/p1689640271500029

**Comment by FFFiend** [comment]
Some notes for myself: 

It is possible for there to be window states with no action events associated with them, i.e. 
we can have 
```
recording_window_events = get_window_events(recording)
recording_action_events = get_action_events(recording)
len(recording_window_events) != len(recording_action_events)
```

**Comment by abrichr** [comment]
Please review https://github.com/OpenAdaptAI/OpenAdapt/pull/420 and let's discuss 1-1

**Comment by FFFiend** [comment]
TODO: create gpt 4 class, 
will add more provider classes gradually today

**Comment by abrichr** [comment]
@FFFiend notes from our conversation:

```
[
{ “action”: {“name”: … }, “window”: {“...”: …, }},
{ “action”: {“name”: … }, “window”: {“...”: …, }},
…
]

For GPT: require a prompt and completion

Autoregressive on a single recording?

prompt:
[
{ “action”: {“name”: … }, “window”: {“...”: …, }}_{1},
{ “action”: {“name”: … }, “window”: {“...”: …, }}_{2},
…
{ “action”: {“name”: … }, “window”: {“...”: …, }}_{k}   # number of reference actions should be configurable, goal is to use diffs to remove unnecessary information, in order to maximize number actions that can be in a single prompt for a fixed context window
# this also involves knowing/calculating the context size of a model (this should also be in the API), and subtracting config.NUM_REF_WINDOW_IN_COMPLETION in order to fit the maximum possible number)

]

Completion:
{ “action”: {“name”: … }, “window”: {“...”: …, }}_{k+1},
{ “action”: {“name”: … }, “window”: {“...”: …, }}_{k+2},  # number of window/actions in completion should be configurable, default to 1, config.NUM_REF_WINDOW_IN_COMPLETION
.
Where k is the timestamp of each action, window pairing 


Def finetune(**kwargs):
     


For action, window pairing in processed_recording:
	finetune((action,window))

For reference_window, reference_action, target_window, target_action in get_finetune_pairs(processed_recording):
    finetune(reference_window, reference_action, target_window, target_action)

get_finetune_pairs(processed_recording):
    “””
    Loop through recording actions in order, and yield tuples of (ref_win, ref_act, trg_win, trg_act)
         
    “””

    For prev_action, next_action in zip(processed_recording.actions, processed_recording.actions[1:]):
        Prev_window = prev_action.window
        Next_window = next_action.window

        Ref_win = prev_window
        Ref_act = prev_action
        Trg_win = next_window
        Trg_act = next_action

        Yield ref_win, ref_act, trg_win, trg_act


.
.

**

Llama says you can’t use it for training another model
Freewilly is only for research

Define licenses for models inside the API.
Define  context length for each model inside API.
Class License:
    #APACHE2: “APACHE2”
    #LLAMA2: “LLAMA2”
    COMMERCIAL = “COMMERCIAL”
    NONCOMMERCIAL = “NONCOMMERCIAL”

Class Capability:
    TRAINING
    TUNING
    MODEL_TRAINING
    INFERENCE

ML_MODELS = [
    {
        "name": "gpt-4",
        "provider": openai.ml.openai,
        # "tune_func": openadapt.ml.openai.gpt4.tune,
        # "infer_func": openadapt.ml.openai.gpt4.infer,
        "modalities": openadapt.ml.openai.gpt4.MODALITIES,  #  gpt4/__init__.py: MODALITIES =  [Modality.TEXT],
        "license": License.FOO,
        "capabilities": [Capability.TRAINING, Capability.TUNING, …],
    },
     {
        "name": "BLIP2",
        "provider": openai.ml.huggingface,
        # "tune_func": openadapt.ml.huggingface.blip2.tune,
        # "infer_func": openadapt.ml.huggingface.blip2.infer,
        "modalities": openadapt.ml.huggingface.blip2.MODALITIES,  #  huggingface/__init__.py: MODALITIES =  [Modality.TEXT, Modality.IMAGE],
    },
    ...
]


config.ML_MODEL_LICENSES
```

**Comment by abrichr** [comment]
We need to filter this out:
```
'AXRoleDescription': 'text entry area', 'AXValue': 'Last login: Wed Jul 19 19:20:24 on ttys000\nowaiszahid@Owaiss-MacBook-Pro modelapi % git clone https://github.com/Owais-Enhancement-Org/OpenAdapt.git\nCloning into \'OpenAdapt\'...\nremote: Enumerating objects: 3311, done.\nremote: Counting objects: 100% (1505/1505), done.\nremote: Compressing objects: 100% (345/345), done.\nremote: Total 3311 (delta 1313), reused 1215 (delta 1141), pack-reused 1806\nReceiving objects: 100% (3311/3311), 24.74 MiB | 29.08 MiB/s, done.\nResolving deltas: 100% (2127/2127), done.\nowaiszahid@Owaiss-MacBook-Pro modelapi %
```

**Comment by abrichr** [comment]
We also want to add the `recording.task_description` to the prompts/finetune pairs

**Comment by abrichr** [comment]
Related: https://github.com/OpenAdaptAI/OpenAdapt/issues/430

File: openadapt/ml_models/openai_module/gpt3_davinci/finetune.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                key: val
+                for key, val in utils.row2dict(reference_action, follow=False).items()
+                if val is not None
+                and not key.endswith("timestamp")
+                and not key.endswith("id")
+                #and not isinstance(getattr(models.ActionEvent, key), property)
+            })
+        ]
+
+    # TODO: find way to condense data inside state key, as it eats up context length
```
Comment (FFFiend) [inline openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: @LaPetiteSouris WDYT about the structure of this finetuning pipeline? (Currently implemented only for the GPT 3.5 Davinci model as GPT-4 is not available for finetuning.) 

We finetune autoregressively here but sanitizing the dicts will need some more work as there is a lot of context being eaten up by the meta and data keys of each Window Event.

Comment (LaPetiteSouris) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: This is exactly my pain point when working on #421 . To formalize the problem:

- It is hard to decide which data to go into the evaluation/tuning pipelines, as we don't know which peace of information may work and which not. Also, each piece of info given will change the output of the models.
- I think there has been efforts on trying to "solve" the model output problems by using guardrails, which was not as we expected. 

I am thinking that we should, for now, limit the scope to the pipeline only, without worrying about the data.

How to sanitize data, which data to put in, to get state, windows or not, is _not_ the scope of this issue. Here we try to define the pipelines and let's not worry much about the performance of the model given the data training.

We can even get simple as just using `action` and ignoring for now `window`. As soon as we get this pipeline, we can experiment things all together and come up with an idea of how to represent the data, how to summarize the data into prompts... etc

In short, let's just use `action`, or strip `windows` to its core for the sake of achieving the fine-tuning pipelines.

Thoughts ? @FFFiend @abrichr

Comment (LaPetiteSouris) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: PS: To complete, or extend this `sanitize` or `summarize` method for each model, IMHO, is the responsibility of the folk who pick up the subject of fine-tuning a specific model, not you who are putting the pipelines in place.

Comment (FFFiend) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: In that case, the pipeline depends currently on the providers and models that we wish to use meaning this is blocked by #419 (correct me if I'm wrong), and extending ```ml_models.py``` itself with more useful high level functions. To that end, I thought of creating an ```add_model``` method to the addition of a provider directory to the user's local ```ml_models``` folder with finetune, infer files along with an addition to the ML_MODELS list with all the necessary info (such as max context length, etc etc). 

Thoughts?

Comment (LaPetiteSouris) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: I think for now, let's keep `ML_MODELS` as it is for the sake of simplicity and move forward.

The next steps may be is to represent this in a form of `pydantic` class. I think database can be an option in the future, but right now we may benefit from restricting the scope of this issue.

For the models which need new provider, we can go in with 2 steps:

- Add a new PR on top of this one later to extend `ml_models`
- Run #419 with the new models.

IHMO, it is the responsibility of anyone who picks up fine-tuning or RF learning (me in this case 👍  )  to add a PR on top to extend `ml_models` and `sanitize` method to make use of the required models.

File: openadapt/crud.py [hunk 3]
```diff
@@ -146,6 +146,10 @@ def get_recording(timestamp):
     return db.query(Recording).filter(Recording.timestamp == timestamp).first()
 
 
+def get_recording_by_id(id: int):
+    return db.query(Recording).filter(Recording.id == id).first()
+    
+
 def _get(table, recording_timestamp):
```
Comment (LaPetiteSouris) [inline openadapt/crud.py:153]: WDYT about adding `get_recording_by_task_description` ?

Reason:

- Most of the time, when I work with fine-tuning, I need to point to a specific record and I never know the `id`, while description is more user-friendly as it is a string given by user.

File: openadapt/ml_models/ml_models.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        "provider": openai_module.gpt3_davinci,
+        "finetune": finetune,
+        "infer": infer,
+        "modalities": MODALITIES,
+        "context_length": 4096,
+        "license": License.COMMERCIAL_USE,
+        "capability": [Capability.TRAINING, Capability.TUNING, Capability.USABLE_OUTPUT, Capability.INFERENCE]
+    }
+]
+
```
Comment (LaPetiteSouris) [inline openadapt/ml_models/ml_models.py:45]: As I mention below, I would just float the thoughts:

Should we break down this into `pydantic` class or `dataclass`  as well ?

File: openadapt/ml_models/openai_module/gpt3_davinci/finetune.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+from openadapt import models
+from openadapt.crud import get_action_events, get_window_events
+import fire
+import openai
+from copy import deepcopy
+import utils
+
```
Comment (LaPetiteSouris) [inline openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:7]: Should we have a generic `ModelFineTuner` like this

```python
class ModelFineTuner(ABC):
    def __init__(self, model_name: str):
        self.model_name = model_name

    @abstractmethod
    def get_finetune_pairs(self, recording: Recording):
        pass
    
    @abstractmethod
    def sanitize_prompt(self, prompt):
        pass

    @abstractmethod
    def tune_model(self, prompt_dataset: Any):
        pass

```

And here in `gpt3_davinci`, you can inherit from `ModelFinetuner` and keep the rest of your code.

It would helps a lot for the following folks who need to fine-tune another model and keep our fine-tune pipeline really organised and structured.

WDYT ?

File: openadapt/ml_models/openai_module/gpt3_davinci/finetune.py (hunk 2 above)
Comment (0dm) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: > In short, let's just use `action`, or strip `windows` to its core for the sake of achieving the fine-tuning pipelines.
> 
> 
> 
> Thoughts ? @FFFiend @abrichr 
> 
> 
@LaPetiteSouris 
I agree, I've done a decent amount of work with the window states and can conclude that there's a lot of unnecessary information for each state -- and some good. There's also OS-specific labels (no Apple Foundation classes on Windows) that may make our work less consistent across platforms. We could strip the window states to, at a minimum:
- window title; ex. opening google chrome and going to outlook would result in a title like Outlook - 10 emails
- width & height of the window: to ensure that our replay strategies are independent of screen sizes
- the "top/left" parts of the window rectangle, and maybe the location of the close button (stored as AXValue)

File: openadapt/crud.py (hunk 3 above)
Comment (FFFiend) [reply openadapt/crud.py:153]: I think this is related to #430 , which would require fine tuning another model on recordings with their labels/completions as the respective task being carried out.

File: openadapt/ml_models/openai_module/gpt3_davinci/finetune.py (hunk 2 above)
Comment (FFFiend) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: I was thinking of straight up removing this to begin with, and keeping everything else since this isn't especially used inside StatefulReplayStrategy (only key presses and clicks are relevant)

Thoughts? @0dm

Comment (LaPetiteSouris) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: Let's strip them to bare bone to make an MVP out of it. Also, it helps a lot in keeping the prompt under the 1024 char limit in some cases.

Comment (FFFiend) [reply openadapt/ml_models/openai_module/gpt3_davinci/finetune.py:77]: We could also..summarize the AXValue key instead of each window state 😄 Related to #430 .

File: openadapt/ml_models/ml_models.py (hunk 4 above)
Comment (FFFiend) [reply openadapt/ml_models/ml_models.py]: added a class ```Availability``` to address physical location 👍

File: openadapt/ml_models/huggingface_models/create_modal_image.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    "openai==0.27.2",
+    "nltk==3.6.2",
+    "sentencepiece==0.1.98",
+    "torch==2.0.0",
+    "transformers",
+    "wandb==0.13.4",
+]
+
+
+def create_conda_image(package: str = None, packages: list[str] = None):
```
Comment (FFFiend) [inline openadapt/ml_models/huggingface_models/create_modal_image.py:47]: I was working on a GistingCompletionProvider and thought that a file like this could be a useful addition to the model API, where we can specify what packages we would like the image to contain before running inference on a modal cloud instance. 

It is worth noting that modal purely hosts functions on the cloud, and not classes.

**Comment by FFFiend** [comment]
Potential things to add: Docker image generation file for models that cant be run on CPU?

File: openadapt/ml_models/base_provider.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+from openadapt import models
+from pydantic import BaseModel
+
+class Modality:
+    TEXT = "TEXT"
+    IMAGE = "IMAGE"
+
+class License:
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:8]: Please format with black, there should be two blank lines between classes 🙏

File: openadapt/ml_models/base_provider.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    COMMERCIAL_USE = "COMMERCIAL_USE"
+    NON_COMMERCIAL_USE = "NON_COMMERCIAL_USE"
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:10]: What do you think about removing `_USE`?

File: openadapt/ml_models/base_provider.py (hunk 7 above)
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:8]: How about `Use` instead of `License`? The latter we will eventually want to use to refer to specific licenses, e.g. MIT

File: openadapt/ml_models/base_provider.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+class Capability:
+    TRAINING = "TRAINING"
+    TUNING = "TUNING"
+    USABLE_OUTPUT = "USABLE_OUTPUT"
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:15]: I think we want to evaluate programmatically instead of hard coding this

File: openadapt/ml_models/base_provider.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    Modalities: list[Modality]
+    Availabilities: list[Availability]
+
+    def infer(self, prompt: str):
+        raise NotImplementedError
+
+    def finetune(self, prompt: str, completion: str):
+        raise NotImplementedError
+
+    def get_children(self):
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:35]: What do you think about calling this e.g. `get_providers` or similar?

File: openadapt/ml_models/base_provider.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+class CompletionProvider(BaseModel):
+    Name: str
+    Capabilities: list[Capability]
+    Modalities: list[Modality]
+    Availabilities: list[Availability]
+
+    def infer(self, prompt: str):
+        raise NotImplementedError
+
+    def finetune(self, prompt: str, completion: str):
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:32]: What do you think about accepting a list of dicts, e.g. `[{"prompt": str, "completion": str}, ...]`

File: openadapt/ml_models/base_provider.py (hunk 10 above)
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:35]: Should this be @classmethod?

File: openadapt/ml_models/completion_factory.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+
+from base_provider import Modality, CompletionProvider
+class CompletionProviderFactory:
+    def get_for_modality(modality: Modality) -> list[CompletionProvider]:
```
Comment (abrichr) [inline openadapt/ml_models/completion_factory.py:4]: What do you think about moving this to CompletionProvider and making it a `classmethod` (or staticmethod) as well?

File: openadapt/ml_models/base_provider.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+from openadapt import models
```
Comment (abrichr) [inline openadapt/ml_models/base_provider.py:1]: What do you think about renaming the `ml_models` directory to simply `ml`?

File: openadapt/ml_models/huggingface_models/create_modal_image.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
+import modal
```
Comment (abrichr) [inline openadapt/ml_models/huggingface_models/create_modal_image.py:1]: What do you think about moving this `utils.py`?

Also consider creating a `utils` directory and refactoring the contents, and moving this `utils/modal.py`.

File: openadapt/ml_models/openai_module/gpt3_davinci/gpt3provider.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+from base_provider import CompletionProvider, Modality, Capability, Availability
```
Comment (abrichr) [inline openadapt/ml_models/openai_module/gpt3_davinci/gpt3provider.py:1]: What do you think about moving to `openadapt/ml/openai/...?

**Comment by abrichr** [comment]
Thanks @FFFiend ! Can you please review https://github.com/CambioML/pykoi to see if we might be able to re-use some of their primitives?

**Comment by FFFiend** [comment]
TODO: create model config API (quantization, temperature, max token size, etc)

**Comment by FFFiend** [comment]
TODO: add metrics infrastructure for finetuning

File: openadapt/ml/huggingface_models/generic_provider.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        return tokenizer
+
+    def fetch_model(model_name: str):
+        """
+        Fetches and returns a model object from HuggingFace
+        """
+        model = AutoModel.from_pretrained(model_name)
+        return model
+
+    def create_pipeline(model_name: str, task: str, trust_remote_code=False):
```
Comment (FFFiend) [inline openadapt/ml/huggingface_models/generic_provider.py:25]: make private (add underscore)

Comment (FFFiend) [reply openadapt/ml/huggingface_models/generic_provider.py]: Decided to remove this completely, unnecessary for now. 

Updated infer method's docstring to contain relevant info about pipeline params.

File: openadapt/ml/openai/gpt3_davinci/gpt3provider.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    ]
+    Modalities = [Modality.TEXT]
+    Availabilities = [Availability.HOSTED]
+
+    def finetune(self, prompt_completion_pair: list[dict[str]]):
+        # TODO
+        pass
+
+    def infer(self, prompt: str):
+        # TODO add call to local finetuned model for inference
```
Comment (abrichr) [inline openadapt/ml/openai/gpt3_davinci/gpt3provider.py:19]: Thanks @FFFiend ! Can you please implement these stubs?

**Comment by FFFiend** [comment]
Added tests. Multiple files have changed because they didn't adhere to black formatting.

File: tests/openadapt/ml/test_ml.py [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    assert len(inference_output) > 0
+
+
+def test_abstract_completion_provider_factory() -> None:
+    """Completion Provider Factory test."""
+    test_modality = Modality.TEXT
+    providers_list = CompletionProviderFactory.get_for_modality(test_modality)
+
+    assert providers_list
```
Comment (abrichr) [inline tests/openadapt/ml/test_ml.py:47]:
```suggestion
    assert providers_list, providers_list
```

Comment (abrichr) [inline tests/openadapt/ml/test_ml.py:47]:
```suggestion
    assert providers_list, (test_modality, providers_list)
```

**Comment by abrichr** [comment]
@FFFiend it looks like this PR includes several unrelated changes (e.g. alembic/versions/5139d7df38f6_add_recording_task_description.py).

Can you please remove all unrelated code? 🙏

**Comment by FFFiend** [comment]
> @FFFiend it looks like this PR includes several unrelated changes (e.g. alembic/versions/5139d7df38f6_add_recording_task_description.py).
> 
> Can you please remove all unrelated code? 🙏

Those changes are purely stylistic (converting single quotes to double, fixing formatting) and were done using ```pre-commit run --all-files``` to fix black formatting across the entire codebase, my checks wouldn't pass without it.

**Comment by FFFiend** [comment]
> @FFFiend it looks like this PR includes several unrelated changes (e.g. alembic/versions/5139d7df38f6_add_recording_task_description.py).
> 
> Can you please remove all unrelated code? 🙏

@abrichr removed all unrelated code, however provider tests (fixed in #488 ) are failing, so just waiting on that to get merged 👍

**Comment by KrishPatel13** [comment]
@FFFiend #488 is merged now. Please update the branch then continue :-)

**Comment by FFFiend** [comment]
@abrichr Ready to merge.

File: openadapt/ml/base_provider.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+    def get_children() -> Any:
+        """Grab all completion provider classes."""
+        return CompletionProvider.__subclasses__()
+
+
+class CompletionProviderFactory:
+    """CompletionProviderFactory class."""
+
+    def get_for_modality(modality: Modality) -> list[CompletionProvider]:
```
Comment (abrichr) [inline openadapt/ml/base_provider.py:67]: Please add `@staticmethod`

File: openadapt/ml/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,13 @@
+"""Exceptions file for the Completions Provider API."""
+
+
+class ModelNotImplementedError(BaseException):
+    """To be raised when a model is not supported by the infrastructure."""
+
+    def __init__(self, model_name: str) -> None:
+        """Init method for the ModelNotImplementedError class."""
```
Comment (abrichr) [inline openadapt/ml/exceptions.py:8]: Please remove

File: openadapt/ml/exceptions.py [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+        super().__init__(f"{model_name} is not currently supported")
+
+
+class GpuNotAvailableError(BaseException):
```
Comment (abrichr) [inline openadapt/ml/exceptions.py:12]: Please rename to `GPUNotAvailableError `

File: openadapt/ml/huggingface_models/__init__.py [hunk 22]
```diff
@@ -0,0 +1 @@
+"""HuggingFace ml package."""
```
Comment (abrichr) [inline openadapt/ml/huggingface_models/__init__.py:1]: Can we please name the submodule `openadapt.ml.huggingface`?

File: openadapt/ml/huggingface_models/generic_provider.py [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+    """GenericHuggingFaceProvider class."""
+
+    Name: str = "Generic HuggingFace Provider"
+    Capabilities: list[Capability] = [
+        Capability.TRAINING,
+        Capability.TUNING,
+        Capability.INFERENCE,
+    ]
+    Modalities: list[Modality] = [Modality.TEXT]
+    Availabilities: list[Availability] = [Availability.HOSTED]
```
Comment (abrichr) [inline openadapt/ml/huggingface_models/generic_provider.py:23]: Should this be `Availabilities: list[Availability] = [Availability.LOCAL_CPU, Availability.LOCAL_GPU]`?

We would need a separate provider for HOSTED models.

Can you please rename this `LocalHuggingFaceProvider`, and add a TODO above the class to implement one or more `HostedHuggingFaceProvider` to support Inference API and Inference Endpoints at https://huggingface.co/docs/huggingface_hub/guides/inference?

File: openadapt/ml/huggingface_models/generic_provider.py [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+    def infer(
+        self,
+        prompt: str,
+        model_path: str,
+        task_description: str,
+        use_pipeline: bool = True,
```
Comment (abrichr) [inline openadapt/ml/huggingface_models/generic_provider.py:30]: Please remove if unused 🙏 

(see https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

File: openadapt/ml/huggingface_models/generic_provider.py [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+        Capability.INFERENCE,
+    ]
+    Modalities: list[Modality] = [Modality.TEXT]
+    Availabilities: list[Availability] = [Availability.HOSTED]
+
+    def infer(
+        self,
+        prompt: str,
+        model_path: str,
+        task_description: str,
```
Comment (abrichr) [inline openadapt/ml/huggingface_models/generic_provider.py:29]: Please rename to `huggingface_task` to differentiate between https://github.com/OpenAdaptAI/OpenAdapt/blob/main/openadapt/models.py#L46

File: openadapt/ml/open_ai/gpt/gptprovider.py [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+        Capability.TUNING,
+        Capability.INFERENCE,
+    ]
+    Modalities: list[Modality] = [Modality.TEXT]
+    Availabilities: list[Availability] = [Availability.HOSTED]
+
+    def finetune(self, prompt_completion_pair: list[dict[str]]) -> None:
+        """Fine-tune the GPT model."""
+        # waiting on FineTuning PR which is currently
+        # a work in progress.
```
Comment (abrichr) [inline openadapt/ml/open_ai/gpt/gptprovider.py:28]: Please modify this comment to say:

```
TODO: implement once fine tuning is merged
```

File: openadapt/ml/open_ai/gpt/gptprovider.py [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+    def infer(self, gpt_model_name: str, prompt: str) -> str:
+        """Users can only infer from models available within their OpenAI organization.
+
+        This includes base models provided by OpenAI as well as
+        finetuned models.
+        """
+        valid_inference_models = openai.Model.list()["data"]
+        for model_dict in valid_inference_models:
+            if model_dict["id"] == gpt_model_name:
+                if gpt_model_name == "gpt-4" or gpt_model_name == "gpt-3.5-turbo":
```
Comment (abrichr) [inline openadapt/ml/open_ai/gpt/gptprovider.py:40]:
```
if gpt_model_name in CHAT_MODEL_NAMES:
```

File: openadapt/ml_utils/modal.py [hunk 28]
```diff
@@ -0,0 +1,47 @@
+"""Conda Image creation tool."""
```
Comment (abrichr) [inline openadapt/ml_utils/modal.py:1]: Can you please move this to `openadapt/ml/utils/modal.py`?

File: tests/openadapt/ml/test_ml.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
+"""Module to test the CompletionProvider API."""
+from openai.error import AuthenticationError
+import pytest
+
+from openadapt.ml.base_provider import CompletionProviderFactory, Modality
+from openadapt.ml.huggingface_models.generic_provider import GenericHuggingFaceProvider
+from openadapt.ml.open_ai.gpt.gptprovider import GPTCompletionProvider
+
+test_gpt_provider = GPTCompletionProvider()
```
Comment (abrichr) [inline tests/openadapt/ml/test_ml.py:9]: Can you please instantiate these inside of a `@pytest.fixture`?

File: openadapt/ml/base_provider.py [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+        return CompletionProvider.__subclasses__()
+
+
+class CompletionProviderFactory:
+    """CompletionProviderFactory class."""
+
+    @staticmethod
+    def get_for_modality(modality: Modality) -> list[CompletionProvider]:
+        """Gets all available CompletionProviders with the given modality."""
+        completion_providers = CompletionProvider.get_children()
```
Comment (FFFiend) [inline openadapt/ml/base_provider.py:70]: @abrichr On adding ```self``` to the ```CompletionProvider``` class, one would have to pass in parameters to the ```CompletionProvider``` that adhere to its attributes types here inside the factory (else it throws the following error):
```
E   pydantic.error_wrappers.ValidationError: 4 validation errors for CompletionProvider
E   Name
E     field required (type=value_error.missing)
E   Capabilities
E     field required (type=value_error.missing)
E   Modalities
E     field required (type=value_error.missing)
E   Availabilities
E     field required (type=value_error.missing)

pydantic/main.py:341: ValidationError
```

 so we would be generating a ```CompletionProvider```` object with specific parameters. Not sure if that's adheres to: https://www.digitalocean.com/community/tutorials/factory-design-pattern-in-java

**Comment by abrichr** [comment]
This is now stale and needs to be updated to support images.

Related: https://github.com/OpenAdaptAI/OpenAdapt/pull/527

@FFFiend do you have any bandwidth? 🙏

**Comment by abrichr** [comment]
@FFFiend would love your thoughts on https://github.com/OpenAdaptAI/OpenAdapt/issues/534 🙏 😄

**Comment by abrichr** [comment]
Closing as part of the architecture transition. The legacy codebase has been moved to `legacy/` and frozen at v0.46.0.

If this change is still needed, please:
1. Open a PR against the relevant modular package repository
2. Or re-open this PR targeting the `legacy/` directory if it's a critical legacy fix

See PR #960 for details on the new meta-package architecture.
